### ISSUE TITLE
Deploy with `build: false` doesn't work

### ISSUE THREAD

**Issue by Bielik20** [serverless-nextjs/serverless-next.js#1162]
### Describe the bug

Running `sls` with build vs running `sls` with manual build do not yield the same result.

This is my config WITH specified build:
```yml
# serverless.yml
distDir: ''

my-app:
  component: "@sls-next/serverless-component@1.19"
  inputs:
    nextConfigDir: ${distDir}
    build:
      cmd: "nx"
      args: ["run", "desktop:build:production"]
      cwd: "../../"
```

This is config WITHOUT build:

```yml
# serverless.yml
distDir: ''

my-app:
  component: "@sls-next/serverless-component@1.19"
  inputs:
    nextConfigDir: ${distDir}
    build: false
```

### Actual behavior

When I run that `nx run desktop:build:production` I do not get `.serverless_nextjs` folder in my dist. Then when I run `sls` to deploy I get:
```
Error: ENOENT: no such file or directory, open 'absolute-path-to-dist/.serverless_nextjs/default-lambda/manifest.js
``` 

### Expected behavior

Running `sls` to deploy with `build: false` after manually building should result in correct deploy.

### Steps to reproduce

Just follow bug description.

### Versions

- OS/Environment: macOS Big Sur 11.2.3
- @sls-next/serverless-component version: 1.19
- Next.js version: 10.1.2

### Additional context

I have searched through the source code and found out that `build` operation actually passes that folder name to `Builder`: https://github.com/serverless-nextjs/serverless-next.js/blob/41b6be1943bca70a1e2e6f919b5ee6fecc56389f/packages/serverless-components/nextjs-component/src/component.ts#L219

Then it is latter used as `buildDir` https://github.com/serverless-nextjs/serverless-next.js/blob/41b6be1943bca70a1e2e6f919b5ee6fecc56389f/packages/libs/lambda-at-edge/src/build.ts#L72

This is all I could find. There is a lot going on that file, and I am not that proficient in serverless DevOps stuff to understand it.

### Checklist

- [x] You have reviewed the README and [FAQs](https://github.com/serverless-nextjs/serverless-next.js#faq), which answers several common questions.
- [x] Please first try using the most recent `latest` or `alpha` `@sls-next/serverless-component` [release version](https://github.com/serverless-nextjs/serverless-next.js/releases), which may have already fixed your issue or implemented the feature you are trying to use. Note that the old `serverless-next.js` component and the `serverless-next.js` plugin are deprecated and no longer maintained.

**Comment by Reiiya** [comment]
I also split build and deploy in different steps and `serverless build --component myAppName works for me. also on 1.19

**Comment by dphang** [comment]
Not sure what you are doing in your command `nx run desktop:build:production` but you can solve it in a few ways:

1. Run the serverless build command as mentioned above
2. You can invoke the builder itself which will build assets into `.serverless_nextjs` directory, this is from `@sls-next/lambda-at-edge` package. Example of how it's done for CDK: https://serverless-nextjs.com/docs/cdkconstruct/
3. There is also `deploy` false configuration which you can use to build without deploying, but then you'd need multiple `serverless.yml` if you also want to deploy (I guess using build and deploy command is easier)

**Comment by Bielik20** [comment]
I am using monorepo. `nx run desktop:build:production` is just a command to build production version of NextJS application.

This is the output I get when running `sls` WITH build configuration:
![image](https://user-images.githubusercontent.com/13436218/121644717-a9a38000-ca93-11eb-8a93-e77cd66dc54c.png)

This is the output I get when running `nx run desktop:build:production` (so the exact the same command that is run in previous step):
![image](https://user-images.githubusercontent.com/13436218/121645260-46661d80-ca94-11eb-9793-dcf24f21cf5a.png)

---

1. Now, I do not want to go with 1, because I have some flags to pass to the build (one of the reasons I wanted to do it outside `sls`).
2. Interesting but I think it will suffer from the same issue. I would have to proxy flags through `serverless.yml`
3. The same issue.

Thank you for all the suggestion. I actually ended up creating a script that runs deploy and I proxy config through `serverless.yml`:

```yml
outputPath: ${env.outputPath}
buildCwd: ${env.buildCwd}
buildTarget: ${env.buildTarget}
customAlias: ${env.customAlias}
customStage: ${env.customStage}
name: ${env.name}
domain: ${env.domain}
subdomain: ${env.subdomain}

web:
  component: '@sls-next/serverless-component@1.19'
  inputs:
    name: ${name}
    bucketName: ${name}
    domain: ['${subdomain}', '${domain}']
    nextConfigDir: ${outputPath}
    runtime: nodejs14.x
    build:
      cmd: 'nx'
      args: ['run', '${buildTarget}', '--customAlias', '${customAlias}', '--customStage', '${customStage}']
      cwd: ${buildCwd}
```

I wish running deploy with `sls` with build set to false would handle the output and add that `.serverless_nextjs` directory on its own. I mean that is something I would expect. If I pass `npm run build` into build command and it works. Then I pass `false` into build command but run exact same command `npm run build` before deploy. I would expect both of those scenarios to work in the same way.

**Comment by Bielik20** [comment]
Hmm, my problem is that I cannot generate `.serverless_nextjs` without running `sls` WITH build command set in `serverless.yml`. 

If I set that build command in `serverless.yml` it works, no problems.

But if I do something like that:

```yml
# serverless.yml
distDir: ''

my-app:
  component: "@sls-next/serverless-component@1.19"
  inputs:
    nextConfigDir: ${distDir}
    build: false
```

And I run:
1. `nx run desktop:build:production` - it builds nextjs app, no problems here.
2. `sls` it fails because `.serverless_nextjs` doesn't exist.

My problem is that `.serverless_nextjs` is NOT the result of `nx run desktop:build:production` (which is simply nextjs build command). It is only created by running `sls` WITH build command set in `serverless.yml`.

So what I find unintuitive, is that setting build command in `serverless.yml` does result in **not just** that command being run, but **extra steps** take place to generate `.serverless_nextjs`.

**Comment by dphang** [comment]
> So what I find unintuitive, is that setting build command in serverless.yml does result in not just that command being run, but extra steps take place to generate .serverless_nextjs.

Yes I think that is expected since the entire lifecycle is 1) Next.js build + 2) component build (generating .serverless_nextjs) and 3) deploying it

Passing a build command is meant to replace just the Next.js build part, maybe we need to update the docs. But doing a component build, you are still building the whole component which includes .serverless_nextjs.

I don't believe there's a way to run just the Next.js build on serverless.yml, as you can just execute that build without this component?

Sorry, I just wasn't too clear what was trying to be achieved...

**Comment by Bielik20** [comment]
Hmm, you nicely described steps. `2)` is what I am missing (that **extra steps**).
Passing no-op command is actually a simple workaround, thanks I haven't thought of it.

Yeah, I think updating docs would be useful, especially in [inputs table](https://github.com/serverless-nextjs/serverless-next.js#inputs). I feel like this is a go to place for people separating build and deploy steps. I am not fan of adding too much configuration but I think it would be nice to have a way of skipping only `1)` in a more descriptive manner than no-op command.

Anyway, thank you for your help. It is way more clear now 🙇

**Comment by dphang** [comment]
Sure, I will update the docs and I think we should be able to close this if there are no further issues.

**Comment by Bielik20** [comment]
@dphang I am sad to say it doesn't work for me. I have got something like:

```yml
# serverless.yml
distDir: ''

my-app:
  component: "@sls-next/serverless-component@1.19"
  inputs:
    nextConfigDir: ${distDir}
    build:
       cmd: 'true'
```

And it fails with:
```
 Error: Cannot find module '/absolute-path-to-dist/.next/routes-manifest.json'
```

It is there after build command. But deploy creates some files and then fails with above error. I tried with version 1.19 and 1.20.0-alpha.23.

**Comment by dphang** [comment]
I think the error seems clear, you don't have a .next directory or it's incompletely. Are you running your Next build command first (separately) and ensuring it is outputting .next directory to the right place (where serverless.yml is)? I think you can use build.cwd as well to change the directory

**Comment by Bielik20** [comment]
I will check that tomorrow at work, but I am doing everything as expected. I took out a build command from `serverless.yml` by replacing it with `true` (as suggested) and I run build before running `sls`.

Running `sls` results in:
```
 Error: Cannot find module '/absolute-path-to-dist/.next/routes-manifest.json'
```
but it is there after running build and before running `sls`.

**Comment by dphang** [comment]
Hm, I am not sure then, do you know which line it is failing at in this component's code (besides the error message). If you are willing you can try to run with `serverless --debug` and/or try to clone this repo and debug which line it is failing at. Or if you can share a minimal repro, I can try to debug myself

**Comment by Bielik20** [comment]
This is what I am getting:
```
0s › Components › Running .

  0s › Template › Deploying .
  DEBUG ─ Resolving the template's static variables.
  DEBUG ─ Collecting components from the template.
  DEBUG ─ Downloading any NPM components found in the template.

  0s › Template › Deploying ..

  0s › Template › Deploying ..

  0s › Template › Deploying ...

  0s › Template › Deploying ...
  DEBUG ─ Analyzing the template's components dependencies.
  DEBUG ─ Creating the template's components graph.
  DEBUG ─ Syncing template state.
  DEBUG ─ Executing the template's components graph.

  0s › Template › Deploying

  1s › nextjs › Deploying

  1s › nextjs › Deploying

error:
Error: Cannot find module '/Users/user-name/Projects/ecommerce/dist/apps/ddb/web/.next/routes-manifest.json'
Require stack:
- /Users/user-name/.serverless/components/registry/npm/@sls-next/serverless-component@1.19.0/node_modules/@sls-next/lambda-at-edge/dist/build.js
- /Users/user-name/.serverless/components/registry/npm/@sls-next/serverless-component@1.19.0/node_modules/@sls-next/lambda-at-edge/dist/index.js
- /Users/user-name/.serverless/components/registry/npm/@sls-next/serverless-component@1.19.0/node_modules/@sls-next/serverless-component/dist/component.js
- /Users/user-name/.serverless/components/registry/npm/@sls-next/serverless-component@1.19.0/node_modules/@sls-next/serverless-component/serverless.js
- /Users/user-name/.nvm/versions/node/v14.16.1/lib/node_modules/serverless/node_modules/@serverless/core/src/Component.js
- /Users/user-name/.nvm/versions/node/v14.16.1/lib/node_modules/serverless/node_modules/@serverless/core/src/index.js
- /Users/user-name/.nvm/versions/node/v14.16.1/lib/node_modules/serverless/node_modules/@serverless/cli/src/index.js
- /Users/user-name/.nvm/versions/node/v14.16.1/lib/node_modules/serverless/bin/serverless.js
  at Function.Module._resolveFilename (internal/modules/cjs/loader.js:880:15)
  at Function.Module._load (internal/modules/cjs/loader.js:725:27)
  at Module.require (internal/modules/cjs/loader.js:952:19)
  at require (internal/modules/cjs/helpers.js:88:18)
  at Builder.<anonymous> (/Users/user-name/.serverless/components/registry/npm/@sls-next/serverless-component@1.19.0/node_modules/@sls-next/lambda-at-edge/dist/build.js:645:36)
  at Generator.next (<anonymous>)
  at fulfilled (/Users/user-name/.serverless/components/registry/npm/@sls-next/serverless-component@1.19.0/node_modules/@sls-next/lambda-at-edge/dist/build.js:5:58) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [
  '/Users/user-name/.serverless/components/registry/npm/@sls-next/serverless-component@1.19.0/node_modules/@sls-next/lambda-at-edge/dist/build.js',
  '/Users/user-name/.serverless/components/registry/npm/@sls-next/serverless-component@1.19.0/node_modules/@sls-next/lambda-at-edge/dist/index.js',
  '/Users/user-name/.serverless/components/registry/npm/@sls-next/serverless-component@1.19.0/node_modules/@sls-next/serverless-component/dist/component.js',
  '/Users/user-name/.serverless/components/registry/npm/@sls-next/serverless-component@1.19.0/node_modules/@sls-next/serverless-component/serverless.js',
  '/Users/user-name/.nvm/versions/node/v14.16.1/lib/node_modules/serverless/node_modules/@serverless/core/src/Component.js',
  '/Users/user-name/.nvm/versions/node/v14.16.1/lib/node_modules/serverless/node_modules/@serverless/core/src/index.js',
  '/Users/user-name/.nvm/versions/node/v14.16.1/lib/node_modules/serverless/node_modules/@serverless/cli/src/index.js',
  '/Users/user-name/.nvm/versions/node/v14.16.1/lib/node_modules/serverless/bin/serverless.js'
  ]
  }
```

I would like to help get that working, but sadly I won't have more time this sprint to dig into that issue. I am executing build as a part of deployment. It has its own problems but we can live with them for now.

**Comment by tavvy** [comment]
@Bielik20 I am also wrestling with getting this to work with `nx`.

I've managed to get a successful deploy with a mixture of the below:
- use `NOW_BUILDER=true` env variable when building as this will create a build using the `{target: 'experimental-serverless-trace'}` flag: https://github.com/nrwl/nx/blob/88019305532e7a2b44c404296048bbbcf6dec2bc/packages/next/plugins/with-nx.ts#L32
- ensure your `serverless.yml` is in the project root when running `serverless` to deploy
- you need to rename a subfolder in nx dist build output folder to what serverless-nextjs is expecting (this will fix your most recently reported error) `mv ./dist/apps/<my next app>/.next/server ./dist/apps/<my next app>/.next/serverless`. Im unsure why/when this discrepancy in folder name (`server` or `serverless`) started.
- my `serverless.yml` has this setup. Note that this is to work around a issue where `serverless-nextjs` will empty your `.next` folder before starting a build: https://github.com/serverless-nextjs/serverless-next.js/blob/master/packages/libs/lambda-at-edge/src/build.ts#L666 this is **even** if you use a `no-op` like 'true' as your build command.
```
webNextjsApplication:
  component: '@sls-next/serverless-component@3.1.0-alpha.3'
  inputs:
    # we re-create this by using a NOW_BUILDER=true variable during build
    useServerlessTraceTarget: true
    build:
      # sls-next has a bug where it cleans up the dist folder assuming the build
      # command will re-create it. We use a previous job to generate the build
      # so to work around this we copy our prebuilt dist into project root
      cmd: 'cp'
      args: ['-r', './dist/apps/web/.', '.']
      cwd: ./
      env: {}
```

**Comment by Bielik20** [comment]
Thank you @tavvy for response. Actually the whole process works for me, save for getting build step out from deploy command. But for other people who stumble upon that, yes:

* One needs `{target: 'experimental-serverless-trace'}` in `next.config.js`. I just have it setup in `next.config.js` all the time. Are there any downsides to that?
* `serverless.yml` I think can be anywhere, as long as to provide correct `nextConfigDir` (the one to your dist).
* This rename issue I have not experienced yet, but I will keep an eye for it.
* This is root of the issue, as you have even described:
> sls-next has a **bug** where it cleans up the dist folder assuming the build

@dphang This really looks like a bug on the `serverless-next.js` side.

I am ok with running build in deploy command for now, but would like change it. Would it be a hard fix to remove that "cleaning up"?

**Comment by dphang** [comment]
@Bielik20 ah, that makes sense now, I think it was there before I started on this project, it might be to ensure. But I guess any build commands (whether Next.js's or custom ones) should handle the cleanup? It looks like just a one line removal if so

**Comment by dphang** [comment]
Actually it might be needed for the normal build flow, in order to ensure only the necessary files are included. I added an option to disable cleaning up `.next` dir: https://github.com/serverless-nextjs/serverless-next.js/pull/1273

**Comment by dphang** [comment]
Closing as the above fix should help with this. Feel free to comment in case of any other concerns.

**Comment by Bielik20** [comment]
Yeah, I can confirm it works now. I have tested it with version `3.1.0-alpha.15`.
Is here a stable release that contains those fixes? Previously I used `1.19` because that was the newest I have found in release tab that wasn't alpha.

**Comment by dphang** [comment]
Not yet, I will release it soon. There were a couple regressions in the latest alphas which were fixed so I think it can be released.

Now that e2e tests are a bit more complete/stable I think we can go for a faster release cycle

**Comment by dan-cooke** [comment]
@dphang just FYI. This `cleanupDotEnv` functionality has regressed somewhere along the line.

Raising an issuse now
